src/popupmenu.c

```diff
@@ -709,7 +740,47 @@ pum_redraw(void)
 					MB_PTR_ADV(rt);
 				    } while (cells > pum_width);
 
-				    if (cells < pum_width)
+				    if (need_ellipsis)
+				    {
+					orig_rt = rt;
+					while (*orig_rt != NUL)
+					{
+					    char_cells = has_mbyte ? (*mb_ptr2cells)(orig_rt) : 1;
+					    if (used_cells + char_cells > 3)
+						break;
+					    used_cells += char_cells;
+					    MB_PTR_ADV(orig_rt);
+					    last_char = orig_rt;
+					}
+
+					if (last_char != NULL)
+					{
+					    if (used_cells < ellipsis_width)
+					    {
+						over_cell = ellipsis_width - used_cells;
+						MB_PTR_ADV(orig_rt);
+						last_char = orig_rt;
+					    }
+					    kept_len = STRLEN(last_char);
+					    new_str = alloc(ellipsis_width + over_cell + kept_len + 1);
+					    if (!new_str)
+						return;
```

Comment:
Seems like you also want to use `vim_memset(new_str, ' ', ellipsis_width + over_cell + kept_len);`